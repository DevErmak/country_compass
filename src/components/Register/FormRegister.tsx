import { useEffect, useState } from 'react';
import './form-register.css';
import { SubmitHandler, useForm } from 'react-hook-form';
import { ZodError, isValid, z } from 'zod';
import { zodResolver } from '@hookform/resolvers/zod';
import { setAuthentication } from '../../store/user/infoUserSlice';
import { useDispatch } from 'react-redux';

type Props = {};

export default function FormRegister({}: Props) {
  const dispatch = useDispatch();

  const formSchemaRegister = z
    .object({
      login: z
        .string()
        .trim()
        .min(1, { message: 'Field is required' })
        .email('Email is not correct'),
      password: z
        .string()
        .min(6, { message: 'The password must be long' })
        .regex(new RegExp('^(?=.*[a-z])(?=.*[A-Z])(?=.*\\d)(?=.*).+$'), {
          message:
            'the password must contain at least 6 characters and have at least one capital letter and number',
        }),
      confirmPassword: z.string().min(6, { message: 'The password must be long' }),
    })
    .refine((schema) => (schema.password === schema.confirmPassword ? true : false), {
      message: 'passwords do not match',
      path: ['confirmPassword'],
    });

  type RegisterFields = z.infer<typeof formSchemaRegister>;

  const {
    register,
    handleSubmit,
    reset,
    setFocus,
    trigger,
    formState: { isDirty, isSubmitting, errors, isValid },
  } = useForm<RegisterFields>({
    mode: 'all',
    resolver: zodResolver(formSchemaRegister),
  });

  const onSubmit: SubmitHandler<RegisterFields> = (data) => {
    let userInfo = JSON.parse(localStorage.getItem('userInfo') as string);
    if (userInfo !== null)
      if (userInfo.find((item: any) => item.name === data.login) === undefined) {
        userInfo.push({ name: data.login, password: data.password, listFavorite: [] });
        localStorage.setItem('userInfo', JSON.stringify(userInfo));
        dispatch(setAuthentication(true));
      } else {
        alert('such a user exists');
      }
    else {
      userInfo = [];
      userInfo.push({ name: data.login, password: data.password, listFavorite: [] });
      localStorage.setItem('userInfo', JSON.stringify(userInfo));
      dispatch(setAuthentication(true));
    }

    reset();
  };

  useEffect(() => {
    setFocus('login');
  }, []);

  return (
    <div className="container-register">
      <div className="modal-register">Register</div>
      <form onSubmit={handleSubmit(onSubmit)} className="form-register">
        <div>{errors.login && errors.login?.message}</div>

        <div>
          <label
            htmlFor="username"
            className={errors.login ? 'label-login not-valid' : 'label-login'}
          >
            Login:
          </label>
          <input
            {...register('login')}
            type="text"
            className={errors.login ? 'input-login-register not-valid' : 'input-login-register'}
          />
        </div>
        <div>
          <label
            htmlFor="password"
            className={errors.password ? 'label-password not-valid' : 'label-password'}
          >
            {errors.password ? 'Password: ' + errors.password?.message : 'Password:'}
          </label>
          <input
            {...register('password')}
            type="password"
            className={
              errors.password ? 'input-password-register not-valid' : 'input-password-register'
            }
          />
        </div>
        <div>
          <label
            htmlFor="username"
            className={
              errors.confirmPassword
                ? 'label-confirm-password  not-valid'
                : 'label-confirm-password'
            }
          >
            {errors.confirmPassword
              ? 'Confirm password: ' + errors.confirmPassword?.message
              : 'Confirm password:'}
          </label>
          <input
            {...register('confirmPassword')}
            type="password"
            className={
              errors.confirmPassword
                ? 'input-confirm-password-register not-valid'
                : 'input-confirm-password-register'
            }
          />
        </div>
        <div className="btn-register">
          <button type="submit" id="btn-register" disabled={!isValid}>
            Register
          </button>
        </div>
      </form>
    </div>
  );
}
