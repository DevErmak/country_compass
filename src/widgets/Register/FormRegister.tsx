import { useEffect } from 'react';
import './form-register.css';
import { SubmitHandler, useForm } from 'react-hook-form';
import { zodResolver } from '@hookform/resolvers/zod';
import { useLazyQuery, useMutation } from '@apollo/client';
import { GET_FAVORITECOUNTRIES, REGISTER, SET_FAVORITECOUNTRIES } from '../../shared/api/graphqlV1';

import { Zoom, toast } from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';
import { useCookies } from 'react-cookie';
import { RegisterFields, formSchemaRegister } from './model/RegisterFormSchema';
import Loader from '../../shared/ui/Loader';
import {
  addFavoriteCountry,
  setCountryBuffer,
  setModal,
} from '../../entities/viewer/model/user/infoUserSlice';
import { formModal } from '../../entities/viewer/model/user/types';
import { useDispatch } from 'react-redux';
import { getBufferCountry } from '../../entities/viewer/model/user/userSelectors';
import { useSelector } from 'react-redux';

type Props = {};

export default function FormRegister({}: Props) {
  const isEmptyLodash = require('lodash.isempty');
  const dispatch = useDispatch();

  useEffect(() => {
    setFocus('login');
  }, []);

  // const formSchemaRegister = z
  //   .object({
  //     login: z
  //       .string()
  //       .trim()
  //       .min(1, { message: 'Field is required' })
  //       .email('Email is not correct'),
  //     password: z
  //       .string()
  //       .min(6, { message: 'The password must be long' })
  //       .regex(new RegExp('^(?=.*[a-z])(?=.*[A-Z])(?=.*\\d)(?=.*).+$'), {
  //         message:
  //           'the password must contain at least 6 characters and have at least one capital letter and number',
  //       }),
  //     confirmPassword: z.string().min(6, { message: 'The password must be long' }),
  //   })
  //   .refine((schema) => (schema.password === schema.confirmPassword ? true : false), {
  //     message: 'passwords do not match',
  //     path: ['confirmPassword'],
  //   });

  // type RegisterFields = z.infer<typeof formSchemaRegister>;

  const {
    register,
    handleSubmit,
    reset,
    setFocus,
    formState: { errors, isValid },
  } = useForm<RegisterFields>({
    mode: 'all',
    resolver: zodResolver(formSchemaRegister),
  });

  const [registerUser] = useMutation(REGISTER);

  const fullInfoCountryBuffer = useSelector(getBufferCountry);
  const [cookie, setCookie] = useCookies(['accessToken']);
  // const [setFavoriteCountry] = useMutation(SET_FAVORITECOUNTRIES);
  // const [getFavoriteCountry] = useLazyQuery(GET_FAVORITECOUNTRIES);

  // useEffect(() => {
  //   if (data !== undefined) {
  //     if (data.registerUser !== null) {
  //       if (data.registerUser.__typename === 'AccessToken') {
  //         //console.log('gkj---------------->', data.login.token);
  //         setCookie('accessToken', data.registerUser.token);
  //       } else
  //         toast.error('you were unable to register', {
  //           position: 'bottom-left',
  //           autoClose: 1500,
  //           hideProgressBar: true,
  //           closeOnClick: true,
  //           pauseOnHover: true,
  //           draggable: true,
  //           progress: undefined,
  //           theme: 'colored',
  //           transition: Zoom,
  //         });
  //     }
  //   }
  // }, [data]);

  const onSubmit: SubmitHandler<RegisterFields> = async (data) => {
    const { data: dataToken } = await registerUser({
      variables: {
        createUser: {
          login: data.login,
          password: data.password,
        },
      },
    });
    if (dataToken && !isEmptyLodash(dataToken)) {
      console.log('---------------->token', dataToken.registerUser.token);
      setCookie('accessToken', dataToken.registerUser.token);
      dispatch(setCountryBuffer({}));
      dispatch(setModal({ isActiveModal: false, formModal: formModal.login }));
    }

    // if (error)
    //   toast.error('you were unable to register', {
    //     position: 'bottom-left',
    //     autoClose: 1500,
    //     hideProgressBar: true,
    //     closeOnClick: true,
    //     pauseOnHover: true,
    //     draggable: true,
    //     progress: undefined,
    //     theme: 'colored',
    //     transition: Zoom,
    //   });
    // else if (token && !isEmptyLodash(token)) {
    //   setCookie('accessToken', token.registerUser.token);
    // }

    reset();
  };
  return (
    <div className="container-register">
      <div className="modal-register">Register</div>
      <form onSubmit={handleSubmit(onSubmit)} className="form-register">
        <div className={'error-login-reg '}>{errors.login && errors.login?.message}</div>
        <div>
          <label
            htmlFor="username"
            className={errors.login ? 'label-login not-valid' : 'label-login'}
          >
            Login:
          </label>
          <input
            {...register('login')}
            type="text"
            className={errors.login ? 'input-login-register not-valid' : 'input-login-register'}
          />
        </div>
        <div className={'error-password-reg'}>{errors.password && errors.password?.message}</div>
        <div>
          <label
            htmlFor="password"
            className={errors.password ? 'label-password not-valid' : 'label-password'}
          >
            Password:
          </label>
          <input
            {...register('password')}
            type="password"
            className={
              errors.password ? 'input-password-register not-valid' : 'input-password-register'
            }
          />
        </div>
        <div className={'error-confirm-password'}>
          {errors.confirmPassword && errors.confirmPassword?.message}
        </div>
        <div>
          <label
            htmlFor="username"
            className={
              errors.confirmPassword
                ? 'label-confirm-password  not-valid'
                : 'label-confirm-password'
            }
          >
            Confirm password:
          </label>
          <input
            {...register('confirmPassword')}
            type="password"
            className={
              errors.confirmPassword
                ? 'input-confirm-password-register not-valid'
                : 'input-confirm-password-register'
            }
          />
        </div>
        <div className="btn-register">
          <button type="submit" id="btn-register" disabled={!isValid}>
            Register
          </button>
        </div>
      </form>
    </div>
  );
}
