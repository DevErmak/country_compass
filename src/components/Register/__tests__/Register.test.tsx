import { renderComponentWithStore } from '../../../shared/lib/test-helpers';
import FormRegister from '../FormRegister';

it('renders correctly FormRegister', () => {
  const { container } = renderComponentWithStore({
    Component: FormRegister,
    preloadedState: {
      infoCountries: {
        listCountries: [],
        fullInfoCountry: [],
        isFullInfoCountry: false,
        isLoading: false,
        infoErrorResponse: '',
      },
      infoUser: {
        userName: '',
        listFavoriteCountries: [],
        isAuthentication: false,
        infoModal: { isActiveModal: false, formModal: 'login' },
      },
    },
  });
  expect(container).toMatchSnapshot();
});
