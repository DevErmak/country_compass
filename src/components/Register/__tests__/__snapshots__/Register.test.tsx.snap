// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly FormRegister 1`] = `
<div>
  <div
    class="container-register"
  >
    <div
      class="modal-register"
    >
      Register
    </div>
    <form
      class="form-register"
    >
      <div />
      <div>
        <label
          class="label-login"
          for="username"
        >
          Login:
        </label>
        <input
          class="input-login-register"
          name="login"
          type="text"
        />
      </div>
      <div>
        <label
          class="label-password"
          for="password"
        >
          Password:
        </label>
        <input
          class="input-password-register"
          name="password"
          type="password"
        />
      </div>
      <div>
        <label
          class="label-confirm-password"
          for="username"
        >
          Confirm password:
        </label>
        <input
          class="input-confirm-password-register"
          name="confirmPassword"
          type="password"
        />
      </div>
      <div
        class="btn-register"
      >
        <button
          disabled=""
          id="btn-register"
          type="submit"
        >
          Register
        </button>
      </div>
    </form>
  </div>
  <div
    id="portal"
  />
</div>
`;
